### PR TITLE
Add fast grouping method for CategoricalArray keys

### PR DESCRIPTION
**Pull request by nalimilan**
When grouping on a single `CategoricalArray` column, we don't need a `Dict` to get group codes.
This approach cannot be used when joining, as the `Dict` is needed.
`group_rows` now handles the sorting, which allows skipping this step if the grouping is guaranteed to be sorted (as is the case for `CategoricalArray`).
This mechanism can be extended in the future to handle multiple `CategoricalArray` keys, or integer keys (using radix sort).

This makes grouping about 4 times faster on simple benchmarks, even with the optimized code from https://github.com/JuliaData/DataFrames.jl/pull/1565 (which BTW is still used when grouping on several `CategoricalArray` keys). Of course grouping only represents a small part of the time taken by a split-apply-combine operation, so typically the win isn't that big.
```julia
using DataFrames, BenchmarkTools
df = DataFrame(rand(500_000, 10));
df.key = categorical(string.(rand(1:100, 500_000)));

# Master
julia> @btime groupby(df, :key);
  8.258 ms (51 allocations: 15.45 MiB)

# This PR
julia> @btime groupby(df, :key);
  2.284 ms (64 allocations: 7.64 MiB)

df = DataFrame(rand(500_000, 10));
df.key = categorical(string.(rand(1:10_000, 500_000)));

# Master
julia> @btime groupby(df, :key);
  27.451 ms (54 allocations: 15.68 MiB)

# This PR
julia> @btime groupby(df, :key);
  7.595 ms (69 allocations: 7.86 MiB)
```

### CODE REVIEW

File: src/dataframerow/utils.jl [hunk 1]
```diff
@@ -4,11 +4,11 @@
 struct RowGroupDict{T<:AbstractDataFrame}
     "source data table"
     df::T
-    "number of groups"
+    "number of groups (i.e. highest group index: some groups may be empty)"
```
Comment (bkamins) [inline src/dataframerow/utils.jl]: nonempty group?

File: src/dataframerow/utils.jl [hunk 2]
```diff
@@ -70,36 +70,30 @@ function hashrows_col!(h::Vector{UInt},
 end
 
 # Calculate the vector of `df` rows hash values.
-function hashrows(df::AbstractDataFrame, skipmissing::Bool)
-    rhashes = zeros(UInt, nrow(df))
-    missings = fill(false, skipmissing ? nrow(df) : 0)
-    for (i, col) in enumerate(columns(df))
+function hashrows(cols::Tuple{Vararg{AbstractVector}}, skipmissing::Bool)
+    len = length(cols[1])
```
Comment (bkamins) [inline src/dataframerow/utils.jl]: this assumes that `cols` is not `()` which is probably the case - asking to make sure that this corner case is not relevant.

File: src/dataframerow/utils.jl [hunk 3]
```diff
@@ -70,36 +70,30 @@ function hashrows_col!(h::Vector{UInt},
...
-    row_group_slots(ntuple(i -> df[i], ncol(df)), rhashes, missings, groups, skipmissing)
-end
-
 function row_group_slots(cols::Tuple{Vararg{AbstractVector}},
-                         rhashes::AbstractVector{UInt},
-                         missings::AbstractVector{Bool},
+                         hash::Val = Val(true),
                          groups::Union{Vector{Int}, Nothing} = nothing,
-                         skipmissing::Bool = false)
+                         skipmissing::Bool = false)::Tuple{Int, Vector{UInt}, Vector{Int}, Bool, Bool}
```
Comment (bkamins) [inline src/dataframerow/utils.jl]: why is this needed?

File: src/dataframerow/utils.jl [hunk 4]
```diff
@@ -70,36 +70,30 @@ function hashrows_col!(h::Vector{UInt},
...
     @assert groups === nothing || length(groups) == length(cols[1])
-    # inspired by Dict code from base cf. https://github.com/JuliaData/DataFrames.jl/pull/17#discussion_r102481481
+    rhashes, missings = hashrows(cols, skipmissing)
```
Comment (bkamins) [inline src/dataframerow/utils.jl]: why is this needed and was not needed earlier?

File: src/dataframerow/utils.jl [hunk 5]
```diff
@@ -70,36 +70,30 @@ function hashrows_col!(h::Vector{UInt},
...
-                         groups::Union{Vector{Int}, Nothing} = nothing,
-                         skipmissing::Bool = false)
-    rhashes, missings = hashrows(df, skipmissing)
-    row_group_slots(ntuple(i -> df[i], ncol(df)), rhashes, missings, groups, skipmissing)
-end
-
 function row_group_slots(cols::Tuple{Vararg{AbstractVector}},
-                         rhashes::AbstractVector{UInt},
-                         missings::AbstractVector{Bool},
+                         hash::Val = Val(true),
```
Comment (bkamins) [inline src/dataframerow/utils.jl]: I understand that you do not use this value now, but add it for the future. Right?

Comment (bkamins) [reply src/dataframerow/utils.jl]: Or it is to make sure that we can also call this method on a single `CategoricalVector` if we pass `Val{true}`?

File: src/dataframerow/utils.jl [hunk 6]
```diff
@@ -135,14 +129,47 @@ function row_group_slots(cols::Tuple{Vararg{AbstractVector}},
...
+end
+
+function row_group_slots(cols::Tuple{CategoricalVector},
+                         hash::Val{false},
+                         groups::Union{Vector{Int}, Nothing} = nothing,
+                         skipmissing::Bool = false)::Tuple{Int, Vector{UInt}, Vector{Int}, Bool, Bool}
+    col = cols[1]
+    @assert groups === nothing || length(groups) == length(col)
+
+    # If missings are to be skipped, they will all go to group 1
```
Comment (bkamins) [inline src/dataframerow/utils.jl]: and this group will be skipped; otherwise missings to group `ngroups`. Right?

File: src/dataframerow/utils.jl [hunk 7]
```diff
@@ -135,14 +129,47 @@ function row_group_slots(cols::Tuple{Vararg{AbstractVector}},
...
+    return ngroups, UInt[], Int[], true, true
 end
 
 # Builds RowGroupDict for a given DataFrame.
 # Partly uses the code of Wes McKinney's groupsort_indexer in pandas (file: src/groupby.pyx).
-function group_rows(df::AbstractDataFrame, skipmissing::Bool = false)
+# - hash: whether row hashes should be computed (if false, the rhashes and gslots fields
+#   hold empty vectors)
+# - sort: whether groups should be sorted
+# - skipmissing: whether rows with missing values should be skipped rather than put into group 0
```
Comment (bkamins) [inline src/dataframerow/utils.jl]: group `ngroups` not `0` unless I do not understand something.

File: src/dataframerow/utils.jl [hunk 8]
```diff
@@ -170,11 +197,37 @@ function group_rows(df::AbstractDataFrame, skipmissing::Bool = false)
...
     if skipmissing
-        splice!(starts, 1)
-        splice!(stops, 1)
+        popfirst!(starts)
+        popfirst!(stops)
         ngroups -= 1
     end
 
+    # if row_group_slots can return empty groups, remove them
+    if empty
```
Comment (bkamins) [inline src/dataframerow/utils.jl]: is it possible that `empty` is `false`?

File: test/dataframerow.jl [hunk 9]
```diff
@@ -45,7 +45,7 @@ module TestDataFrameRow
     @test hash(DataFrameRow(df, 2)) != hash(DataFrameRow(df, 6))
 
     # check that hashrows() function generates the same hashes as DataFrameRow
-    df_rowhashes, _ = DataFrames.hashrows(df, false)
+    df_rowhashes, _ = DataFrames.hashrows(tuple(DataFrames.columns(df)...), false)
```
Comment (bkamins) [inline test/dataframerow.jl]: I would rather write `Tuple(columns(df))` to avoid splatting and you do not have to quality `columns` with `DataFrames.`.

File: src/dataframerow/utils.jl (hunk 1 above)
Comment (nalimilan) [reply src/dataframerow/utils.jl]: I'm not sure I understand the question. Some of the "groups" counted here may be empty, so that just gives a value guaranteed to be as high or higher than the actual number of groups. (Actually this field doesn't seem to be used, so we could remove it.)

File: src/dataframerow/utils.jl (hunk 2 above)
Comment (nalimilan) [reply src/dataframerow/utils.jl]: Yes, since you can't group on no key. :-)

File: src/dataframerow/utils.jl (hunk 5 above)
Comment (nalimilan) [reply src/dataframerow/utils.jl]: Yes, that's just a way to dispatch to the right method: the optimized `CategoricalArray` method can only be called with a single `CategoricalArray` column and when `hash::Val{false}`; this one is the fallback, and can be called with any columns and `hash` value.

File: src/dataframerow/utils.jl (hunk 3 above)
Comment (nalimilan) [reply src/dataframerow/utils.jl]: Yeah, that's annoying, it's because we call it from `group_rows`, which takes a `DataFrame`, and therefore doesn't know the types of the columns it passes to `row_group_slots`. Since all `row_group_slots` return the same argument types, the compiler should be able to figure the return type anyway, but that doesn't seem to work without the assertions.

File: src/dataframerow/utils.jl (hunk 4 above)
Comment (nalimilan) [reply src/dataframerow/utils.jl]: It was called by the `row_group_slots` helper before, but now we call it only from this method, and not from the optimized `CategoricalArray` one.

File: src/dataframerow/utils.jl (hunk 8 above)
Comment (nalimilan) [reply src/dataframerow/utils.jl]: Woops, yes, that's the case with the fallback method, but for some reason I put `true`...

File: src/dataframerow/utils.jl (hunk 1 above)
Comment (bkamins) [reply src/dataframerow/utils.jl]: I asked if we count only non-empty groups. But if it is not used maybe better remove it?

File: src/dataframerow/utils.jl (hunk 2 above)
Comment (bkamins) [reply src/dataframerow/utils.jl]: Yes, but `groupby(df, [])` throws now an obscure error (`BoundsError`) - maybe some clearer message should be printed.

File: src/dataframerow/utils.jl (hunk 6 above)
Comment (nalimilan) [reply src/dataframerow/utils.jl]: I've expanded the comment a bit as that's confusing.

File: src/dataframerow/utils.jl (hunk 1 above)
Comment (nalimilan) [reply src/dataframerow/utils.jl]: Let's remove it.

File: src/dataframerow/utils.jl (hunk 7 above)
Comment (nalimilan) [reply src/dataframerow/utils.jl]: Actually, the group isn't necessarily the first nor the last one if `sort=false`.

**Comment by bkamins** [comment]
Thank you. Looks good. One thing that might be added is to internally document the meaning of return value of `row_group_slots` somewhere for future developers (in a comment probably is enough).

**Comment by bkamins** [comment]
Thank you. I think we are good to merge.

Do you have any benchmark how we stand now in the benchmark against other frameworks?

**Comment by nalimilan** [comment]
> Do you have any benchmark how we stand now in the benchmark against other frameworks?

Not yet. I'd like to implement the type-stable `combine` here or in DataFramesMeta first.

**Comment by nalimilan** [comment]
Though it shouldn't make a big difference for https://github.com/h2oai/db-benchmark/issues/30 when there are enough rows per group. So we could ask them to update the benchmark and run it again. But `SubDataFrame` deprecations will probably slow down the simplest syntax, so better drop them first.

**Comment by bkamins** [comment]
They rerun them automatically on the latest release of benchmarked packages and they are updated regularly AFAIK, so we simply have to move forward with the package I guess 😄.
